### PR TITLE
AST Dump

### PR DESCRIPTION
**Pull request by KCreate**

### CODE REVIEW

File: lexer/src/main/kotlin/tools/samt/lexer/Lexer.kt [hunk 1]
```diff
@@ -287,7 +287,7 @@ class Lexer private constructor(
         currentPosition.copy(
                 charIndex = currentPosition.charIndex + 1,
                 row = currentPosition.row + 1,
-                col = -1,
+                col = 0,
```
Comment (PascalHonegger) [inline lexer/src/main/kotlin/tools/samt/lexer/Lexer.kt:290]: This might cause a problem given multiline input, the first line tokens are 0-indexed but with this change the \
See the failing unit test for a case which breaks with just this change:
https://github.com/samtkit/core/blob/e03554a2a06c8c27c52f280b286755249a2f0e86/lexer/src/test/kotlin/tools/samt/lexer/LexerTest.kt#L264-L283

File: parser/src/main/kotlin/tools/samt/parser/Nodes.kt [hunk 2]
```diff
@@ -2,7 +2,11 @@ package tools.samt.parser
 
 import tools.samt.common.Location
 
-sealed class Node(val location: Location)
+abstract class Node(val location: Location) {
```
Comment (PascalHonegger) [inline parser/src/main/kotlin/tools/samt/parser/Nodes.kt:5]: Why change `sealed` to `abstract`?

File: parser/src/main/kotlin/tools/samt/parser/ASTPrinter.kt [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+            is ArrayNode -> {
+                node.values.forEach(body)
+            }
+
+            is BundleIdentifierNode -> {
+                node.components.forEach(body)
+            }
+
+            else -> {}
```
Comment (PascalHonegger) [inline parser/src/main/kotlin/tools/samt/parser/ASTPrinter.kt:205]: I stronlgy prefer the `Node` class to be sealed so we can leave out this `else` block and get a compiler error if new nodes are added in the future

File: parser/src/main/kotlin/tools/samt/parser/ASTPrinter.kt [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        }
+    }
+
+    companion object {
+        fun dump(node: Node): String {
+            return ASTPrinter().dump(node)
+        }
```
Comment (PascalHonegger) [inline parser/src/main/kotlin/tools/samt/parser/ASTPrinter.kt:212]: Could be a top-level function instead of having this `ASTPrinter` wrapper class

File: common/src/main/kotlin/tools/samt/common/Location.kt [hunk 5]
```diff
@@ -1,15 +1,21 @@
...
 data class FileOffset(val charIndex: Int, val row: Int, val col: Int) {
-    override fun toString(): String = "$row:$col"
+    override fun toString(): String = "${row + 1}:${col + 1}"
 }
 
 data class Location(
-    val start: FileOffset,
-    val end: FileOffset = start,
+    var start: FileOffset,
+    var end: FileOffset = start,
```
Comment (PascalHonegger) [inline common/src/main/kotlin/tools/samt/common/Location.kt:9]: Why change to `var`?

File: common/src/main/kotlin/tools/samt/common/Location.kt [hunk 6]
```diff
@@ -1,15 +1,21 @@
...
 ) {
     init {
-        require(start.row <= end.row) { "Location end row (${end.row}) cannot be before start row ${start.row}" }
+        if (start.row > end.row) {
+            end = start
+        }
+
+        // FIXME: There is some bug with this enabled
```
Comment (PascalHonegger) [inline common/src/main/kotlin/tools/samt/common/Location.kt:16]: I would like to fix this before merging this, just making this comment as a reminder

File: parser/src/main/kotlin/tools/samt/parser/ASTPrinter.kt [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        is IdentifierNode -> node.name
+        is ImportBundleIdentifierNode -> if (node.isWildcard) "wildcard" else null
+        is IntegerNode -> node.value.toString()
+        is FloatNode -> node.value.toString()
+        is BooleanNode -> node.value.toString()
+        is StringNode -> "\"${node.value}\""
+        else -> null
+    }
+
+    private inline fun iterateChildren(node: Node, body: (Node) -> Unit) {
```
Comment (PascalHonegger) [inline parser/src/main/kotlin/tools/samt/parser/ASTPrinter.kt:60]: Such a parameter would usually be called `block` instead of `body` as far as I know

File: cli.sh
Comment (PascalHonegger) [inline cli.sh:1]: It's a bit black-magic that the `cli.sh` script runs a build first, but that's probably okay to test within the repository

But for consistency sake, a `./cli.bat` for windows and a `./cli` without the `.sh` would be the correct approach (like `gradle` does)

**Review by PascalHonegger — CHANGES_REQUESTED** [review]

File: parser/src/main/kotlin/tools/samt/parser/Nodes.kt [hunk 8]
```diff
@@ -2,7 +2,11 @@ package tools.samt.parser
...
+    override fun toString(): String {
+        return ASTPrinter.dump(this)
```
Comment (PascalHonegger) [inline parser/src/main/kotlin/tools/samt/parser/Nodes.kt:7]: Would it be cleaner if the ASTPrinter would be part of the CLI? If we want to use ANSI-Colors etc. which are CLI-Specific it would make sense to not have a dependency on such things within the Parser-Nodes themselves.

File: cli/src/main/kotlin/tools/samt/cli/CliArgs.kt [hunk 9]
```diff
@@ -12,6 +12,9 @@ class CliArgs {
     @Parameter(names = ["--benchmark-runs"], description = "Parse files multiple times and measure execution time", help = true)
     var benchmarkRuns: Int = 100
 
+    @Parameter(names = ["--dumpast"], description = "Dump a visual representation of the AST", help = true)
```
Comment (PascalHonegger) [inline cli/src/main/kotlin/tools/samt/cli/CliArgs.kt:15]: What about `--ast` instead?

File: cli/src/main/kotlin/tools/samt/cli/CliCompiler.kt [hunk 10]
```diff
@@ -3,11 +3,12 @@ package tools.samt.cli
 import tools.samt.common.DiagnosticConsole
 import tools.samt.common.DiagnosticContext
 import tools.samt.lexer.Lexer
+import tools.samt.parser.FileNode
 import tools.samt.parser.Parser
 import tools.samt.parser.ParserException
 import java.io.File
 
-fun parse(files: List<String>) {
+fun parse(files: List<String>, dumpAST: Boolean = false) {
```
Comment (PascalHonegger) [inline cli/src/main/kotlin/tools/samt/cli/CliCompiler.kt:11]: Given seperation of duties, it would be a lot cleaner to have the `dumpAST` only within the CLI module, we don't really want a side-effect like writing to the standard out within the Parser, this would be dubious given the language server integration in the future

File: common/src/main/kotlin/tools/samt/common/Location.kt (hunk 5 above)
Comment (KCreate) [reply common/src/main/kotlin/tools/samt/common/Location.kt:9]: Temp fix below required it. Will be removed on the final version

File: cli/src/main/kotlin/tools/samt/cli/CliArgs.kt (hunk 9 above)
Comment (PascalHonegger) [reply cli/src/main/kotlin/tools/samt/cli/CliArgs.kt:15]: Fair point, I just somehow don't like the current name 🤔 
`--dump-ast` is more readable but a bit long, but if we want a `--dump-tokens`, `--dump-semantic` or `--dump-types` in the future this might make more sense

Comment (KCreate) [reply cli/src/main/kotlin/tools/samt/cli/CliArgs.kt:15]: That was my reasoning as well, once we add more flags to dump more things we might want to be clear about what is actually happening, instead of just "--semantic", which could mean any different amount of things.

File: lexer/src/main/kotlin/tools/samt/lexer/Lexer.kt [hunk 11]
```diff
@@ -18,25 +18,27 @@ class Lexer private constructor(
         this.reader = if (reader.markSupported()) reader else BufferedReader(reader)
     }
 
-    /**
-     * Index of the last read character, 0 based
-     */
+    // the starting position of the current token window
+    // gets reset after each token is read
```
Comment (PascalHonegger) [inline lexer/src/main/kotlin/tools/samt/lexer/Lexer.kt:22]: I would use KDoc string here (`/** */` instead of `//`)

File: common/src/main/kotlin/tools/samt/common/Location.kt [hunk 12]
```diff
@@ -4,18 +4,13 @@ data class FileOffset(val charIndex: Int, val row: Int, val col: Int) {
     override fun toString(): String = "${row + 1}:${col + 1}"
 }
 
+// start and end file offsets of a location. start is inclusive, end is exclusive.
```
Comment (PascalHonegger) [inline common/src/main/kotlin/tools/samt/common/Location.kt:7]: I would use KDoc string here (`/** */` instead of `//`)

File: cli/src/main/kotlin/tools/samt/cli/CliArgs.kt (hunk 9 above)
Comment (KCreate) [reply cli/src/main/kotlin/tools/samt/cli/CliArgs.kt:15]: I agree, I will change it.

File: parser/src/main/kotlin/tools/samt/parser/ASTPrinter.kt (hunk 3 above)
Comment (KCreate) [reply parser/src/main/kotlin/tools/samt/parser/ASTPrinter.kt]: The problem with including each node class is that a lot of the nodes do not have any child nodes, so all their handler blocks would just include a `do nothing here` comment. I see the advantage of getting a compiler error when adding new nodes though so I'm going to add those handlers anyway.

**Comment by KCreate** [comment]
@PascalHonegger Please re-review and merge if satisfied

**Comment by PascalHonegger** [comment]
@KCreate please rebase on main and then mark PR as ready (you could also rename the `Make Pascal happy` commit in the process)

**Comment by KCreate** [comment]
That commit name cannot be changed because it is relevant to the structural integrity of the repository.

File: cli.sh
Comment (KCreate) [reply cli.sh:1]: It's just used for testing, so I don't have to manually run a build every single time I want to execute the CLI.

File: cli/src/main/kotlin/tools/samt/cli/CliCompiler.kt [hunk 13]
```diff
@@ -3,45 +3,52 @@ package tools.samt.cli
...
-            Parser.parse(tokenStream, diagnostics)
-        } catch (e: ParserException) {
-            diagnostics.printToConsole()
-            continue
-        }
+        add(parsedFileNode!!)
+    }
+}
+
+fun parseSource(filePath: String, fileSource: String, diagnostics: DiagnosticConsole): FileNode? {
```
Comment (PascalHonegger) [inline cli/src/main/kotlin/tools/samt/cli/CliCompiler.kt:41]: If this is intended as an internal helper function it should be made private

File: cli/src/main/kotlin/tools/samt/cli/CliCompiler.kt [hunk 14]
```diff
@@ -3,45 +3,52 @@ package tools.samt.cli
...
+    if (diagnostics.hasErrors()) {
+        return null
+    }
 
-        diagnostics.printToConsole()
-        println(fileNode)
+    return try {
+        Parser.parse(filePath, tokenStream, diagnostics)
+    } catch (e: ParserException) {
+        null
```
Comment (PascalHonegger) [inline cli/src/main/kotlin/tools/samt/cli/CliCompiler.kt:51]: Maybe add a comment like `The parse exception message in also added to diagnostics, we can safely ignore it here`

File: cli/src/test/kotlin/tools/samt/cli/CliCompilerTest.kt [hunk 15]
```diff
@@ -10,4 +17,84 @@ class CliCompilerTest {
             parse(listOf("foo.samt"))
         }
     }
+
+    @Test
+    fun `correctly formats an AST dump`() {
```
Comment (PascalHonegger) [inline cli/src/test/kotlin/tools/samt/cli/CliCompilerTest.kt:22]: Could be a seperate test file for the `ASTPrinter`

File: common/src/main/kotlin/tools/samt/common/Location.kt [hunk 16]
```diff
@@ -1,12 +1,15 @@
...
+    override fun toString(): String = "${row + 1}:${col + 1}"
 }
 
+/**
+ * start and end file offsets of a location. start is inclusive, end is exclusive.
+ * */
 data class Location(
     val start: FileOffset,
-    val end: FileOffset = start,
+    val end: FileOffset = start
```
Comment (PascalHonegger) [inline common/src/main/kotlin/tools/samt/common/Location.kt:12]: My lovely trailing comma 😢

File: specification/examples/debug.samt
Comment (PascalHonegger) [inline specification/examples/debug.samt:1]: Intentionally added?

File: specification/examples/greeter.samt [hunk 17]
```diff
@@ -15,14 +15,18 @@ record GreetResponse {
...
   timestamp: DateTime
 }
 
 @Author("Leonard Schütz")
-@Description("
+@Description("foo bar
   This is some very long comment
   which describes the service.
-")
+baz")
```
Comment (PascalHonegger) [inline specification/examples/greeter.samt:29]: I personally like it more if the `greeter.samt` stays "realistic", but we could edit / move the `debug.samt` to include all kinds of edge-cases

File: cli/src/main/kotlin/tools/samt/cli/CliCompiler.kt (hunk 13 above)
Comment (KCreate) [reply cli/src/main/kotlin/tools/samt/cli/CliCompiler.kt:41]: It is used in the ASTPrinter unit test. Ideally I'm going to refactor the whole CLI / diagnostics / compiler interactions anyway so this function won't live very long.

File: common/src/main/kotlin/tools/samt/common/Location.kt (hunk 16 above)
Comment (KCreate) [reply common/src/main/kotlin/tools/samt/common/Location.kt:12]: Will be back soon

File: specification/examples/debug.samt
Comment (KCreate) [reply specification/examples/debug.samt:1]: Just a file that can be used to test things during development. Doesn't really need to be committed but I don't see a reason to remove it either. I'll put it into the gitignore if you like.

File: specification/examples/greeter.samt (hunk 17 above)
Comment (KCreate) [reply specification/examples/greeter.samt:29]: I agree

File: cli.sh
Comment (PascalHonegger) [reply cli.sh:1]: Given our cross-platform approach I still prefer my obove mentioned approach, but we'll have to adjust the dev-workflow once we create a end-user and developer documentation so it's fine for now
